sdk/attestation/azure-security-attestation/src/attestation_client.cpp

```diff
@@ -177,14 +183,31 @@ namespace {
 std::shared_timed_mutex SharedStateLock;
 }
 
-std::vector<AttestationSigner> const& AttestationClient::GetAttestationSigners(
+/**
+ * @brief Retrieves the information needed to validate the response returned from the attestation
+ * service.
+ *
+ * @details Validating the response returned by the attestation service requires a set of
+ * possible signers for the attestation token.
+ *
+ * @param context Client context for the request to the service.
+ */
+void AttestationClient::RetrieveResponseValidationCollateral(
     Azure::Core::Context const& context) const
 {
   std::unique_lock<std::shared_timed_mutex> stateLock(SharedStateLock);
 
   if (m_attestationSigners.size() == 0)
   {
-    m_attestationSigners = GetAttestationSigningCertificates(context).Value.Signers;
+    auto request
+        = AttestationCommonRequest::CreateRequest(m_endpoint, HttpMethod::Get, {"certs"}, nullptr);
+    auto response = AttestationCommonRequest::SendRequest(*m_pipeline, request, context);
```

Comment:
There's a potential risk for a nasty bug around this code.
Usually, a critical path should not perform an I/O. 

Consider re-factoring this method to lock the mutex once you get the I/O response and after you parse it.
I think it would be better to have multiple threads sending the I/O and letting the first one to come back successfully to set the m_attestationSigners. Instead of relaying in only one thread doing the I/O (if something goes wrong with it, all other threads are blocked)

Reply:
I suspect that you're also requesting that I not hold the lock when I call `m_attestationSigners.size() == 0`.  And that drives me straight into the arms of the double checked lock pattern, which is considered unsafe to use without some fancy use of `std::atomic`.

Could you please explain why this code is unsafe or what the bug is?

Also, what makes this a "critical path"? Why is this more critical than any other client method?

Or is the suggestion that we should ALWAYS make a network call in this method, even if it's unnecessary? That seems rather inefficient.

Reply:
by `critical path` I was referring to a `Critical Section` 😄  A place which depends on a mutex lock/unlock.

The code is not unsafe, but it could lead a multi-thread program to get stuck -> and program to be irresponsive.
If the first request which locks the Mutex goes to an irresponsive state (by the nature of the I/O call), and if there's not a cancellation token, then, any other thread trying to send a request will get stuck waiting for the mutex to be released.

I will suggest to follow the same pattern we have for `refresh token`.  I think @JeffreyRichter  and @antkmsft  expended many hours on it and even published a patent out of it. 
You could even consider moving this method as part of an attestation `policy` (similar to the authentication policy) and make the `m_attestationSigners` be a field of the policy (similar to a auth token).
Then the policy check the field on each request and when it is needed, it calls `RetrieveResponseValidationCollateral` (similar to calling `refreshToken()`).  // Using a policy, you wouldn't need all the ASSERT calls to the methods.

I know it is funny. A magic operation, calling 2 service APIs is not allowed. But an operation using a policy to make an extra call is currently accepted (that's the auth policy).  So, it might be OK to explore it as an alternative.

Anyway, I am not planning on blocking this PR on this detail. I will let it to your consideration.

It's been said in the past, that a class constructor must never do an I/O operation.  I would expect the same thing within a critical section.

Reply:
When Jeff, you, and I discussed this issue, the idea of implementing this using a pipeline policy was considered, but there is no functional difference between a pipeline policy creating a second I/O operation and a client method creating a second I/O operation - there is still an I/O operation happening that cannot be accounted for in AKS bookkeeping.  And the RetrieveResponseValidationCollateral mechanism is orders of magnitude easier to implement.

I cannot find either "refresh token" or "refreshtoken" by searching through the source code, so I'm not sure exactly what this "refresh token" technique that Anton and Jeff created is.

What I think I can do safely is to acquire the lock and use it solely to protect the m_attestationSigners variable. That avoids the DCLP because the m_attestationSigners variable will never be accessed outside the lock.

It is still going to have the problem you mentioned above of multiple clients hanging on an operation, but that operation is always a network operation.

Reply:
Look at:
  sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp:429
   Defines the `mutable Credentials::AccessToken m_accessToken;`

   sdk/core/azure-core/src/http/bearer_token_authentication_policy.cpp:24

Set the token or refresh it:
```cpp
{
    std::lock_guard<std::mutex> lock(m_accessTokenMutex);

    // Refresh the token in 2 or less minutes before the actual expiration.
    if (std::chrono::system_clock::now() > (m_accessToken.ExpiresOn - std::chrono::minutes(2)))
    {
      m_accessToken = m_credential->GetToken(m_tokenRequestContext, context);
    }

    request.SetHeader("authorization", "Bearer " + m_accessToken.Token);
  }
```

I don't fully understand it. But it looks to me that the GetToken is taking the lock and will do the I/O to refresh...  So, we might be doing this already :(   

@JeffreyRichter , as part of the Refresh token, if multiple clients (say 100 clients), call an operation at the same time:
   - Are all the client relying on the first one who takes the lock to get a token?  
   - What happen if that client hangs?

Reply:
I know what happens with the token authentication policy. I'm saying that there is no functional difference between hiding a REST API call in a policy and hiding it in a service method.

I'm making a huge assumption that the reason that the AKS folks care that there is at most one network I/O for every API call is because they care about capacity monitoring and auto-scaleout and having more than one network I/O for every API call messes up their capacity planning metrics (because this is the only reason I could come up why they would care about this particular issue - network and CPU capacity matters a great deal to automatic load balancing services like AKS). 

If that is the case, then hiding a network I/O in a policy vs hiding it in a service method is a distinction without difference, since both of them add an additional network operation.

Hiding the network call in a policy complicates the service code and just hides the network operation in a location which is hard to find. I'd rather make things explicit.

Reply:
FWIW, the pipeline code above is making a service call while holding an exclusive lock, which has the exact same "multiple threads hung up on an uncancellable network call" problem you mentioned above. For identity, I'm not sure there's a reasonable way around that issue (because I'm not sure what happens if the same bearer token is retrieved multiple times). For attestation, we can safely release the lock and acquire multiple times because the collateral being retrieved is effectively static (the same collateral is returned for each of the multiple threads)